### PR TITLE
Package and command for diffing Kubernetes objects

### PR DESCRIPTION
**Pull request by squaremo**
This adds some machinery for diffing generic objects, then specialises that to the Kubernetes objects that we typically deal with, then use that to implement a `fluxctl diff` command.

The package `diff` provides generic representations of diffs, and some algorithms for diffing values of various kinds. There is a specialisation hook: implementing the interface `Differ` for a type means that will get used instead of the generic diff algorithm.

The package `platform/kubernetes/diff` contains some of the types used in Kubeternetes config, and some procedures for loading them from a file or directory, and some specialisations via `Differ` (e.g., for secrets). 

The idea here is to build the types out of the fields that matter when applying -- then it can be used to see if, for example, a commit contains a "meaningful" change, and therefore whether to apply it to the cluster. I have probably not captured all such fields; it may take a bit of trial and error to figure out what we care about.

Lastly, this adds a `fluxctl diff` command, which just checks one file (or directory) against another. This is mainly just to prove the use of the packages, but may eventually be useful in itself.

### CODE REVIEW

File: cmd/fluxctl/diff_cmd.go [hunk 1]
```diff
@@ -0,0 +1,88 @@
...
+	if len(args) != 2 {
+		return newUsageError("please supply two filenames")
+	}
+
+	output := func(diffs diff.ObjectSetDiff) error {
+		diffs.Summarise(cmd.OutOrStdout())
+		return nil
+	}
+
+	switch opts.format {
```
Comment (squaremo) [inline cmd/fluxctl/diff_cmd.go]: This is a bit pointless; I mainly wanted it for seeing the raw (`printf(%s)`) output for debugging.

File: diff/diff.go [hunk 2]
```diff
@@ -0,0 +1,293 @@
...
+// Difference represents a "chunk" of difference between two
+// `Object`s.
+type Difference interface {
+	Summarise(out io.Writer)
+	AtPath() string
+}
+
+type Chunk struct {
+	Deleted []interface{}
+	Added   []interface{}
```
Comment (philwinder) [inline diff/diff.go]: To have better type safety, could the added/deleted be strings? Or io.Readers?
Do they need to be interfaces?

File: diff/diff.go [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+	"reflect"
+	"sort"
+	"strings"
+)
+
+// Difference represents a "chunk" of difference between two
+// `Object`s.
+type Difference interface {
+	Summarise(out io.Writer)
+	AtPath() string
```
Comment (philwinder) [inline diff/diff.go]: Does a difference always have a path? 
E.g. if there is no file in the repository then there won't be a path, right?

File: diff/diff.go [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+
+type Object interface {
+	ID() ObjectID
+	Source() string
+}
+
+// ObjectSet is a set of several objects which can be diffed
+// collectively.
+type ObjectSet struct {
+	Source  string
```
Comment (philwinder) [inline diff/diff.go]: What is Source?

File: diff/diff.go [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+func MakeObjectSet(source string) *ObjectSet {
+	return &ObjectSet{
+		Source:  source,
+		Objects: map[ObjectID]Object{},
+	}
+}
+
+type ObjectSetDiff struct {
+	A, B      *ObjectSet
+	OnlyA     []Object
```
Comment (philwinder) [inline diff/diff.go]: Why do we need OnlyA/B? Isn't that information contained within the Differences.Added?

File: diff/diff.go [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+type OpaqueChunk struct {
+	Path string
+}
+
+func (c OpaqueChunk) AtPath() string {
+	return c.Path
+}
+
+// Objects are considered unique by {Namespace, Kind, Name}
+type ObjectID struct {
```
Comment (philwinder) [inline diff/diff.go]: General question. What happens if the user alters the namespace or name in the repository (unlikely to happen in the cluster)?
Then all elements in the repo will be marked as repo only, and the same for the cluster, but cluster only.
Probably fine.

File: diff/diff.go (hunk 2 above)
Comment (philwinder) [inline diff/diff.go]: Yeah AFAICT they are always `[]Difference` types.

File: diff/diff.go [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+	}
+	return diffValue(reflect.ValueOf(a), reflect.ValueOf(b), typA, "")
+}
+
+var differType = reflect.TypeOf((*Differ)(nil)).Elem()
+
+// Compare two reflected values and compile a list of differences
+// between them.
+func diffValue(a, b reflect.Value, typ reflect.Type, path string) ([]Difference, error) {
+	if typ.Implements(differType) {
```
Comment (philwinder) [inline diff/diff.go]: Could all the diff*'s implement the differType? Then you wouldn't need the switch below?

File: diff/diff.go [hunk 8]
```diff
@@ -0,0 +1,293 @@
...
+			diffs = append(diffs, fieldDiffs...)
+		}
+	}
+	return diffs, nil
+}
+
+// diff each element, and include over- or underbite. TODO report an
+// array of ground values as a single chunk, rather than recursing.
+func diffArrayOrSlice(a, b reflect.Value, sliceTyp reflect.Type, path string) ([]Difference, error) {
+	var changed []Difference
```
Comment (philwinder) [inline diff/diff.go]: Grim. Well done, but grim. I wonder if there's a library to do this for you?

File: diff/diff_test.go [hunk 9]
```diff
@@ -0,0 +1,194 @@
...
+
+	expected := []Difference{
+		Added("shamu", "map[four]"),
+		Removed("baz", "map[three]"),
+		Changed("bar", "bart", "map[two]"),
+	}
+	if !reflect.DeepEqual(expected, diffs) {
+		t.Errorf("expected diff:\n%#v\ngot:\n%#v\n", expected, diffs)
+	}
+}
```
Comment (philwinder) [inline diff/diff_test.go]: 💯

File: platform/kubernetes/resource/load.go [hunk 10]
```diff
@@ -0,0 +1,98 @@
...
+		}
+	}
+	if err := chunks.Err(); err != nil {
+		return objs, fmt.Errorf(`scanning multidoc from "%s": %s`, source, err.Error())
+	}
+	return objs, nil
+}
+
+// ---
+// Taken directly from https://github.com/kubernetes/apimachinery/blob/master/pkg/util/yaml/decoder.go.
```
Comment (philwinder) [inline platform/kubernetes/resource/load.go]: Can't you reuse that package then?

File: platform/kubernetes/resource/spec.go [hunk 11]
```diff
@@ -0,0 +1,106 @@
+package resource
+
+import (
+	"fmt"
+
+	"github.com/weaveworks/flux/diff"
+)
+
+// Types that daemonsets, deployments, and other things have in
```
Comment (philwinder) [inline platform/kubernetes/resource/spec.go]: I'm worried that we're coupling ourselves to the k8s API here. I.e. we can't support older clusters (e.g. they don't have annotations) and we'd have to update if they add anything.

Comment (philwinder) [inline platform/kubernetes/resource/spec.go]: And also we have the tedious task of going through each K8s type and adding it to this list.

**Comment by squaremo** [comment]
Taking this off the radar (but leaving the branch for now)

**Pull request closed without merging**